internal/access/credential.go

```diff
@@ -10,153 +10,134 @@ import (
 
 	"github.com/infrahq/infra/internal"
 	"github.com/infrahq/infra/internal/generate"
-	"github.com/infrahq/infra/internal/logging"
 	"github.com/infrahq/infra/internal/server/data"
 	"github.com/infrahq/infra/internal/server/models"
 	"github.com/infrahq/infra/internal/validate"
 )
 
-func CreateCredential(c *gin.Context, user models.Identity) (string, error) {
-	db, err := RequireInfraRole(c, models.InfraAdminRole)
+func CreateCredential(c *gin.Context, user *models.Identity) (string, error) {
+	tx, err := RequireInfraRole(c, models.InfraAdminRole)
 	if err != nil {
 		return "", HandleAuthErr(err, "user", "create", models.InfraAdminRole)
 	}
 
-	tmpPassword, err := generate.CryptoRandom(12, generate.CharsetPassword)
+	password, err := generate.CryptoRandom(12, generate.CharsetPassword)
 	if err != nil {
-		return "", fmt.Errorf("generate: %w", err)
+		return "", fmt.Errorf("crypto random: %w", err)
 	}
 
-	hash, err := bcrypt.GenerateFromPassword([]byte(tmpPassword), bcrypt.DefaultCost)
-	if err != nil {
-		return "", fmt.Errorf("hash: %w", err)
-	}
-
-	userCredential := &models.Credential{
-		IdentityID:      user.ID,
-		PasswordHash:    hash,
+	credential := &models.Credential{
 		OneTimePassword: true,
 	}
 
-	if err := data.CreateCredential(db, userCredential); err != nil {
+	if err := createCredential(tx, user, credential, password); err != nil {
 		return "", err
 	}
 
-	_, err = data.CreateProviderUser(db, data.InfraProvider(db), &user)
+	return password, err
+}
+
+func createCredential(tx *data.Transaction, user *models.Identity, credential *models.Credential, password string) error {
```

Comment:
A lot of this function feels like it should be happening in the update user handler before the access check to set values in the database, if the structure of the code allows for that.